quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -160,671 +164,714 @@ fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
-            let new = simplify_infix(&new_left, InfixOperator::Slash, b, limit.saturating_sub(1));
-            min_by_key(original, new, size)
-        }
-
-        // Right distribution
-        (
-            ref a,
-            InfixOperator::Star,
-            ref right @ Expression::Infix(InfixExpression {
-                left: ref b,
-                operator: InfixOperator::Plus,
-                right: ref c,
-            }),
-        ) => {
-            let original = mul!(a.clone(), right.clone());
-            let ab = simplify_infix(a, InfixOperator::Star, b, limit.saturating_sub(1));
-            let ac = simplify_infix(a, InfixOperator::Star, c, limit.saturating_sub(1));
-            let new = simplify_infix(&ab, InfixOperator::Plus, &ac, limit.saturating_sub(1));
-            min_by_key(original, new, size)
-        }
-
-        // Left distribution
-        (
-            ref left @ Expression::Infix(InfixExpression {
-                left: ref a,
-                operator: InfixOperator::Plus,
-                right: ref b,
-            }),
-            InfixOperator::Star,
-            ref c,
-        ) => {
-            let original = mul!(left.clone(), c.clone());
-            let ac = simplify_infix(a, InfixOperator::Star, c, limit.saturating_sub(1));
-            let bc = simplify_infix(b, InfixOperator::Star, c, limit.saturating_sub(1));
-            let new = simplify_infix(&ac, InfixOperator::Plus, &bc, limit.saturating_sub(1));
-            min_by_key(original, new, size)
-        }
-
-        //----------------------------------------------------------------
-        // Finally: other parenthesis manipulation
-        //----------------------------------------------------------------
-
-        // Mul inside Div on left, multiplicand = denominator
-        (
-            Expression::Infix(InfixExpression {
-                left: ref multiplier,
-                operator: InfixOperator::Star,
-                right: ref multiplicand,
-            }),
-            InfixOperator::Slash,
-            ref denominator,
-        ) if **multiplicand == *denominator => *multiplier.clone(),
-
-        // Mul inside Div on left
-        (
-            ref numerator @ Expression::Infix(InfixExpression {
-                left: ref multiplier,
-                operator: InfixOperator::Star,
-                right: ref multiplicand,
-            }),
-            InfixOperator::Slash,
-            ref denominator,
-        ) => {
-            let original = div!(numerator.clone(), denominator.clone());
-            let new_multiplicand = simplify_infix(
-                multiplicand,
-                InfixOperator::Slash,
-                denominator,
-                limit.saturating_sub(1),
-            );
-            let new = simplify_infix(
-                multiplier,
+            // Addition associative, right: a + (b + c) = (a + b) + c, pick the shorter
```

Comment:
Could merge this with the next case

Reply:
Same comment as before re: reconstructing the original